### PR TITLE
Add fields to Email Struct. Handle Mailgun 'event-data'

### PR DESCRIPTION
**Pull request by cpursley**
@maartenvanvliet

I took the liberty (hope that's alright) of adding message_id, event, timestamp and raw params to the Email Struct. These fields are shared by both Mandrill and Mailgun. Also, I expanded upon Mailgun to handle [event-data](https://documentation.mailgun.com/en/latest/api-webhooks.html#webhooks) webhook types.

You may want to manually test a bit. I did expand the Mailgun tests, but these could still be beefier. So far, these Mailgun changes are working well for me in production

This will need a version bump before release (if you accept the PR). Let me know if you'd like me to make any modification.

### CODE REVIEW

File: lib/receivex/adapters/mailgun.ex [hunk 1]
```diff
@@ -21,35 +23,123 @@ defmodule Receivex.Adapter.Mailgun do
...
+      html: html,
+      text: text,
+      timestamp: Parsers.to_datetime(timestamp),
+      raw_params: email
+    }
   end
 
+  def normalize_params(_), do: nil
+
+  defp from(from), do: parse_address(from)
```
Comment (maartenvanvliet) [inline lib/receivex/adapters/mailgun.ex:141]: should the parse_address and related also be moved to the Parsers module?

File: lib/receivex/email.ex [hunk 2]
```diff
@@ -5,12 +5,27 @@ defmodule Receivex.Email do
...
+          message_id: String.t(),
+          event: String.t(),
           sender: String.t(),
           to: [address],
           from: address,
+          subject: String.t(),
           html: String.t(),
           text: String.t(),
-          subject: String.t()
+          timestamp: DateTime.t(),
```
Comment (maartenvanvliet) [inline lib/receivex/email.ex:16]:
```suggestion
          timestamp: DateTime.t() | nil,
```

File: lib/receivex/parsers.ex [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    case DateTime.from_unix(timestamp) do
+      {:ok, timestamp} ->
+        timestamp
+
+      _ ->
+        nil
+    end
+  end
+
+  def to_datetime(timestamp) when is_float(timestamp) do
```
Comment (maartenvanvliet) [inline lib/receivex/parsers.ex:13]: do any of the api's return float timestamps? 

I feel we can be more assertive here and raise when it doesn't conform to a timestamp, and stick to either strings or integers.

**Review by maartenvanvliet — CHANGES_REQUESTED** [review]
Thanks for this!! Couple of small nits and it can be merged :)

If you have more ideas, let me know!

**Comment by cpursley** [comment]
Thank for the feedback, I will make those changes.

Yeah, Mailgun sends over a float - at least for `event-data` webhook. So maybe it's best to handle timestamp as a string and let the end user deal with it as they want (and this allows more flexibility).

I will make a mention about the event-data webhook types. I'm actually not sure what the previous/existing Mailgun webhook type is from - if you know, can you point met to the docs?

File: lib/receivex/adapters/mailgun.ex [hunk 4]
```diff
@@ -21,35 +23,123 @@ defmodule Receivex.Adapter.Mailgun do
...
-    signature =
-      :crypto.mac(:hmac, :sha256, api_key, data)
-      |> Base.encode16(case: :lower)
+       )
+       when is_map(signature) do
+    valid_signature?(timestamp, token, expected_signature, api_key)
+  end
 
-    Plug.Crypto.secure_compare(signature, expected_signature)
+  defp valid_webhook_request?(
```
[the remark this answers is not in the document]
Comment (cpursley) [reply lib/receivex/adapters/mailgun.ex:40]: Yep, hope that is alright. Will do on comment.

**Comment by cpursley** [comment]
I need to do a bit more testing before this gets merged.

**Comment by cpursley** [comment]
Okay, this is working pretty well for me.

One catch with the Mailgun event-data webhook is many of the Email Struct values are `nil`. So if you think that is too much divergence from a general-purpose structure - maybe we create different Structs per adapter.

**Review by maartenvanvliet — APPROVED** [review]
thanks! ❤️❤️

I'll see if I can whip up a release in the coming days

**Pull request merged**